### PR TITLE
Add JProfiling Documentation

### PR DESCRIPTION
**Pull request by ncough**
This adds some documentation describing the JProfiling
implementation and behaviour.

### CODE REVIEW

File: doc/compiler/JProfiling.md [hunk 1]
```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
...
+As the table populates based on the first distinct values that are
+seen, its possible to have ignored highly frequent values, for example due to
+an early phase change. To account for this, a thread inspects the hash
+tables and clears any with high *other* counters relative to that of the
+profiled values. If it decides that highly frequency values have been
+missed, it will reset the table and allow it to repopulate.
+
+Each table has a lock bit, which is locked when accessing data from a
+compilation thread, rearranging in the helper or clearing in the
+thread. Increments from the mainline do check this lock, to avoid
```
Comment (andrewcraik) [inline doc/compiler/JProfiling.md]: should this be 'do' or 'do not'?

File: doc/compiler/JProfiling.md [hunk 2]
```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
...
+the overhead. Additionally, these increments are not atomic, due to
+the overhead, which may result in races. In testing, this did not affect
```
Comment (andrewcraik) [inline doc/compiler/JProfiling.md]: perhaps replace from overhead to the end of the sentence with 'overhead caused by synchronization'?

File: doc/compiler/JProfiling.md [hunk 3]
```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
...
+results sufficiently to negatively impact optimizations.
+It should be noted that, unlike the global monitor used by other profiling
+implementations, the lock is not reentrant.
+
+In addition to the `pext` hash function, other are supported. One approach
```
Comment (andrewcraik) [inline doc/compiler/JProfiling.md]: other => others

File: doc/compiler/JProfiling.md [hunk 4]
```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
...
+uses an array of bytes specifying shifts to apply to the original value.
+This is implemented using existing opcodes and should be supported on all
+platforms. A more efficient implementation exists using the new `bitpermute`
+opcode, configured with an array of bytes specifying bit indices, rather
+than shifts. If supported by the codegen, the `bitpermute` approach will
```
Comment (andrewcraik) [inline doc/compiler/JProfiling.md]: perhaps name the query which must return true from the cg for the bitpermute to be used?

File: doc/compiler/JProfiling.md [hunk 5]
```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
...
+be used.
+
+## Configuration
+
+Value JProfiling can be used to replace JitProfiling. In this configuration,
+all existing requests for value profiling trees will insert placeholder
+JProfiling helper calls. The JProfilingValue optimization pass will run
+late in the opt strategy, lower these placeholders into the partial IL and
```
Comment (andrewcraik) [inline doc/compiler/JProfiling.md]: strategy and will lower

File: doc/compiler/JProfiling.md [hunk 6]
```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
...
+helper implementation. This can be enabled with `enableJProfilingInProfilingCompilations`.
+Late swaps to profiling are not supported in this mode and will result in
+the compilation restarting.
```
Comment (andrewcraik) [inline doc/compiler/JProfiling.md]: a compilation abort and restart.

**Comment by andrewcraik** [comment]
No build necessary documentation only - a few editing suggestions, but otherwise looks good. Will look to merge once the comments are addressed.

**Comment by ncough** [comment]
Thanks for the suggestions. The documentation has been updated.

**Comment by andrewcraik** [comment]
LGTM - provides a high-level view of the value JProfiling design - no code changes so merging.

**Pull request merged**